walkthrough/05-aws/04-reusing-stacks/README.md

```diff
@@ -0,0 +1,179 @@
+
+```bash
+$ terragrunt run-all destroy --terragrunt-non-interactive
+16:28:49.121 INFO   The stack at /Users/yousif/repos/src/github.com/yhakbar/terragrunt-3313-stacks-walkthrough/walkthrough/05-aws/04-reusing-stacks/live will be processed in the following order for command destroy:
+Group 1
+- Module dev/.terragrunt-stack/stateful-service/.terragrunt-stack/services/api
+- Module prod/.terragrunt-stack/stateful-service/.terragrunt-stack/services/api
+
+Group 2
+- Module dev/.terragrunt-stack/stateful-service/.terragrunt-stack/storage/db
+- Module prod/.terragrunt-stack/stateful-service/.terragrunt-stack/storage/db
+```
+
+At first glance, this is _less DRY_ than the previous walkthrough, as there is a new `stacks/stateful-service/terragrunt.stack.hcl` file that has to be tracked. However, the advantage of this approach is that the `stacks/stateful-service/terragrunt.stack.hcl` file can be re-used across multiple environments, and can be edited in one place to alter the behavior of all environments.
+
+Say, for example, you wanted to add another API to the `stateful-service` Stack. You could add it to the `stacks/stateful-service/terragrunt.stack.hcl` file, and any Stack that references it would automatically have the new API added to it.
+
+Edit the [terragrunt.stack.hcl](./stacks/stateful-service/terragrunt.stack.hcl) file to look like the following:
+
+```hcl
+unit "api" {
+	source = "../../../../units/api"
+	path   = "services/api"
+}
+
+unit "canary_api" {
+    source = "../../../../units/api"
+    path   = "services/canary-api"
+}
+
+unit "db" {
+	source = "../../../../units/db"
+	path   = "storage/db"
+}
+```
+
+Now, if we were to simply regenerate the Stacks, and apply them, we would actually get some errors from AWS. The reason for this is that the logic for the `api` Unit doesn't account for the fact that it might not be the only API in the Stack.
+
+You can see the relevant logic for the `name` input here:
+
+```hcl
+...
+	name = "${local.project_name}-api-${local.random_id}-${local.environment}"
+...
+```
+
+Multiple API units would try to create resources using names like `stacks-walkthrough-05-04-api-zgggh-dev`.
+
+This is a limitation of the design of RFC [#3313](https://github.com/gruntwork-io/terragrunt/issues/3313), and is something that additional tooling might be added in the future to make this more convenient to handle.
+
+The design philosophy of the proposal is that the authors Unit `terragrunt.hcl` files should decide how they want to handle these sorts of conflicts, as they are authoring the configurations that will be passed directly as inputs to OpenTofu/Terraform modules.
+
+In this case, the `path` gives us a useful way of disambiguating the resources that are created by the `api` Unit. We can update the `name` input to reference the name of the directory it is placed in.
+
+```hcl
+...
+	name = "${local.project_name}-${basename(get_terragrunt_dir())}-${local.random_id}-${local.environment}"
+...
+```
```

Comment:
The idea that we use the inputs of the _unit_ for the inputs of the stack is not at all expected, but it's actually really cool. I love the idea that you could define the unit in such a way that the path sets the input value. 

But...it's hard to read and understand, so I'd still really like to see us introduce a "base case" approach to passing inputs to units where you can see inputs specified inline. We could then highlight the limitations of this and call out that users could choose from different options with different pros/cons.

At the end of the day, it's just really jarring to see an instance of a stack (a `terragrunt.stack.hcl` file), and see _no  configuration whatsoever_. The fact that the stack instance and the stack instance configuration are separated or inferred is really surprising, so I'm worried that both Stack authors and Stack consumers (people who just drop into the codebase) are going to be confused by what's happening.

Reply:
Stacks don't have inputs. The single responsibility of a Stack is to define how Units (or other Stacks) fit together as part of a collective.

I think this is an X Y problem. The thing users want isn't actually to set inputs on Stacks. It's to have Stacks with dynamic Units.

I know we've debated the pros and cons of this quite a bit, but I would like to make two final arguments for why I think we should **start** with this approach:

1. This pattern is what users _currently have in their codebases_. Terragrunt users currently don't have a way to parameterize `terragrunt.hcl` files, so they already have logic like this to dynamically determine inputs based on pulling in external configurations, using the names of directories, etc. 

   Having this be how you make Stacks dynamic on release as the "base case" is how we maximize the likelihood that zero code changes to `terragrunt.hcl` files have to occur in order to adopt Stacks (note that in [this walkthrough](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/tree/main/walkthrough/05-aws/02-stacks) I explicitly pointed out that I didn't edit the `terragrunt.hcl` files at all under `This chapter was authored by:`).

   In the [next chapter](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/tree/main/walkthrough/05-aws/03-no-shared), the only way `terragrunt.hcl` files are edited is to _simplify_ them using configurations that already exist in Terragrunt, and work today. No new syntax, and _less_ indirection.

2. This walkthrough wouldn't work with inputs on `terragrunt.stack.hcl` files. 
   
    Note that all of the examples in this repository actually work using the latest release of Terragrunt. The only "magic" functionality used here is [the bash script that generates the Stack](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/blob/main/scripts/mock-stack-generate.sh). 

    While I could just tell readers of the walkthrough to use their imaginations, the practical takeaway of that fact is things like:

    a. Introducing Stacks involves less moving parts in the codebase (no additional logic has to be introduced to have inputs merged between those defined on `terragrunt.hcl` files and `terragrunt.stack.hcl` files via some sort of override mechanism.

    b. The CI systems that users have built will require no updates. Users can adopt Stacks with zero changes to their CI systems, as from an operational perspective, Stacks are transparent. Units work just like they always have, and users can leverage diffs, etc to determine which Units need updates.

OK, it was kind of four arguments, but hopefully this is persuasive that we are fine to move forward with the RFC as is.

Reply:
Hm, this point I disagree on. 

> The thing users want isn't actually to set inputs on Stacks. It's to have Stacks with dynamic Units.

Strong disagree. Consider Brian's canonical use case from GruntCon: deploy a common set of units, each with slightly different configuration. At the end of the day, there is a need for both a _stack definition_ and _stack instance_, and each stack instance needs the potential for unique _configuration_. In my mind, having some solution for each of these is table stakes for this feature.

At an absolute minimum, we need to demonstrate how we handle each of those concepts with the proposed implementation.

> Terragrunt users currently don't have a way to parameterize terragrunt.hcl files

In my mind the Tofu module is the "unit definition" and the `terragrunt.hcl` file is the "unit instance", and you specify `inputs` to give configuration. 

I don't have more time to dig into this right now, but I'd be grateful if you could address the above points.

Reply:
Is it clear that the same `terragrunt.stack.hcl` file is the definition of a stack, and the instance of a stack in different contexts?

In this example, we're using this file as the "definition" of a stack:
https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/blob/main/walkthrough/05-aws/04-reusing-stacks/stacks/stateful-service/terragrunt.stack.hcl

When a stack references it:
https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/blob/main/walkthrough/05-aws/04-reusing-stacks/live/dev/terragrunt.stack.hcl

It generates a `.terragrunt-stack/stateful-service` directory containing that `terragrunt.stack.hcl` file, which then recursively generates more `.terragrunt-stack` directories until all Stacks have been generated.

The `terragrunt.stack.hcl` file in the `.terragrunt-stack/stateful-service` directory is now an "instance" of the stack, with the same contents.

The only file that's committed to the repo is the "definition", but having the file work both ways is what allows Stacks to be recursively nested.

Each instance does have the potential for unique configuration, they just don't do it via `inputs`.

I tried to explain this [here](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/tree/main/walkthrough/05-aws/04-reusing-stacks#stack-dynamicity). If you could, please let me know what part of that doesn't address the concerns here.